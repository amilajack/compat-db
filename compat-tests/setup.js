// @flow
import { join } from 'path';
import { writeFileSync } from 'fs';
import Providers from '../src/providers/Providers';
import PrepareDatabase from '../src/database/PrepareDatabase';
import JobQueue from '../src/database/JobQueueDatabase';
import writeAllJobsToJSON from './jobs';
import { browserNameToCaniuseMappings } from '../src/helpers/Constants';
import {
  convertCaniuseToBrowserName,
  filterDuplicateTargets,
  allTargets,
  getVersionsToMark } from '../src/helpers/GenerateVersions';


const records = Providers();

/**
 * Create 'capabilities' by writing to the temporary file `capabilities.json`.
 *
 * For every job in the JobQueue, remove duplicate targets and format the results
 * to capabilities. `wdio.js` will read the `capabilities.json` file
 */
async function writeCapabilities() {
  const queue = new JobQueue();

  // If there are jobs in the queue already, skip the following steps
  if ((await queue.count()) > 0) return;

  const caniuseIds = (Object.values(browserNameToCaniuseMappings): Array<string>); // eslint-disable-line
  const jobs = [];

  // If no jobs are in the JobQueue, create a job for every record and caniuseId
  records.forEach(record => {
    caniuseIds.forEach((caniuseId) => {
      jobs.push({
        name: caniuseId,
        browserName: convertCaniuseToBrowserName(caniuseId),
        record: JSON.stringify(record),
        version: getVersionsToMark([], caniuseId).middle,
        protoChainId: record.protoChainId,
        type: record.type,
        caniuseId
      });
    });
  });

  await queue.insertBulk(jobs);

  // Take all the targets in JobQueue. Filter duplicates
  const targets = filterDuplicateTargets(
    (await queue.getAll())
      .filter(e =>
        !!e.browserName &&
        !!e.version
      )
      .map(job => ({
        browserName: job.browserName,
        version: job.version,
        platform: allTargets.find(e =>
          e.browserName === job.browserName &&
          e.version === job.version
        ).platform  // eslint-disable-line
      }))
  );

  writeFileSync(
    join(__dirname, '..', 'capabilities.json'),
    JSON.stringify(targets)
  );

  await writeAllJobsToJSON();

  process.exit(0);
}

PrepareDatabase(records, join(__dirname, '..', 'lib', 'all.json'));
writeCapabilities();
