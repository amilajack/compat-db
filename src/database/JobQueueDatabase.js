/**
 * `JobQueue` is responsible for queueing up new tests to be run again on
 * @flow
 */
import AbstractDatabase from './AbstractDatabase';


export type JobQueueType = {
  name: string,
  browserName: string,
  protoChainId: string,
  platform: string,
  version: string,
  record: string,
  type: 'js-api' | 'css-api' | 'html-api',
  caniuseId: string,
};

type whereClauseType = {
  browserName: string,
  protoChainId?: string
}
& (
  { caniuseId: string }
  | { browserName: string }
);

export default class JobQueue extends AbstractDatabase {
  constructor(tableName: string = 'jobs') {
    super(tableName);
  }

  migrate() {
    return super.migrate((table) => {
      table.increments('id').primary();
      table.string('name');
      table.string('browserName');
      table.string('platform');
      table.string('protoChainId');
      table.string('version');
      table.enu('type', ['js-api', 'css-api', 'html-api']);
      table.string('record', 1000);
      table.string('caniuseId');
      table.enu('status', ['queued', 'running']).defaultTo('queued');
    });
  }

  markJobsStatus(whereClause: whereClauseType, status: 'running' | 'queued') {
    return this.connection
      .Database
      .where(whereClause)
      .save({ status }, { method: 'update', patch: true });
  }

  /**
   * Override paramater type with stricter 'Array<JobQueueType>'
   */
  insertBulk(recordsToInsert: Array<JobQueueType>) {
    return super.insertBulk(recordsToInsert);
  }

  remove(whereStatement: JobQueueType) {
    return super.remove(whereStatement);
  }

  /**
   * Find all the compatibility records for every version of the same browser
   */
  find(whereClause: whereClauseType): Promise<Array<JobQueueType>> {
    return this.connection.Database.where(whereClause)
      .fetchAll()
      .then(records => records.toJSON());
  }
}
