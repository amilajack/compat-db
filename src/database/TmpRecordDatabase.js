/**
 * A temporary in-memory database used to store records returned from saucelabs.
 * @flow
 */
import AbstractDatabase from './AbstractDatabase';
import type { RecordType as RType } from '../providers/RecordType';


type str = string;

export type TmpRecordDatabaseRecordType = {
  name: str,
  versions: {
    [version: str]: 'y' | 'n' | 'n/a'
  }
};

type sameRecordType = Promise<TmpRecordDatabaseRecordType>;

type rQueryType = {
  type: 'js-api' | 'css-api' | 'html-api',
  protoChainId: str,
};

export default class TmpRecordDatabase extends AbstractDatabase {
  constructor(tableName: str = 'tmp-records') {
    super(tableName);
  }

  migrate() {
    return super.migrate(table => {
      table.increments('id');
      table.string('name');
      table.string('protoChainId');
      table.string('versions', 1000);
      table.enu('type', ['js-api', 'css-api', 'html-api']);
      table.string('caniuseId');
    });
  }

  /**
   * Find all the compatibility records for every version of the same browser
   */
  findSameVersionCompatRecord(record: rQueryType, caniuseId: str): sameRecordType {
    return this.connection.Database
      .where({
        name: caniuseId,
        type: record.type,
        protoChainId: record.protoChainId,
        caniuseId
      })
      .fetchAll()
      .then(records => records.toJSON())
      .then(records => records.map(_record => ({
        ..._record,
        versions: JSON.parse(_record.versions)
      })))
      .then(_records => (_records.length ? _records[0] : null));
  }

  async insertBulkRecords(record: RType, caniuseId: str, versions: Array<str>, isSupported: bool) {
    const newlyGenerateRecordVersions = {};

    versions.forEach((version) => {
      newlyGenerateRecordVersions[version] = isSupported ? 'y' : 'n';
    });

    return this.connection.Database
      .where({
        caniuseId,
        name: caniuseId,
        protoChainId: record.protoChainId
      })
      .save({
        name: caniuseId,
        type: record.type,
        protoChainId: record.protoChainId,
        versions: JSON.stringify(newlyGenerateRecordVersions),
        caniuseId
      });
  }
}
