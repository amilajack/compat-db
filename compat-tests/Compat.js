// @flow
import 'babel-polyfill';
import dotenv from 'dotenv';
import webdriver from 'selenium-webdriver';
import AssertionFormatter from '../src/assertions/AssertionFormatter';
import {
  insertBulkRecords,
  findSameVersionCompatRecord } from '../src/database/TmpDatabase';
import { getVersionsToMark } from '../src/helpers/GenerateVersions';
import JobQueueDatabase from '../src/database/JobQueueDatabase';
import setup from './setup';
import type { RecordType } from '../src/providers/RecordType';
import type { JobQueueType } from '../src/database/JobQueueDatabase';
import type { browserCapabilityType } from './setup';


/* eslint no-console: 0 */

const jobQueue = new JobQueueDatabase();

dotenv.config();

type finishedTestType = {
  job: JobQueueType,
  record: RecordType,
  isSupported: bool
};
type capabilityType = {
  browserName: string,
  version: string,
  platform: string,
};
type exTestType = Promise<finishedTestType[]>;

/**
 * Take an array of records, generate and run tests, and return the results
 * @TODO: Optimization: Batch tests
 */
export async function executeTests(capability: capabilityType, jobs: JobQueueType[]): exTestType {
  const { browserName, platform, version } = capability;
  const username = process.env.SAUCE_USERNAME;
  const accessKey = process.env.SAUCE_ACCESS_KEY;

  if (!accessKey || !username) {
    throw new Error('Invalid Sauce Labs API key or username');
  }

  const driver = new webdriver
    .Builder()
    .withCapabilities({
      browserName,
      platform,
      version,
      username,
      accessKey
    })
    .usingServer(
      `http://${username}:${accessKey}@ondemand.saucelabs.com:80/wd/hub`
    )
    .build();

  console.log(`Executing ${jobs.length} tests in parallel on ${platform} ${browserName} ${version}`);

  const items = await Promise.all(
    jobs.map(job => {
      const record: RecordType = JSON.parse(job.record);
      const { apiIsSupported } = AssertionFormatter(record);

      return driver
        .executeScript(`return (${apiIsSupported})`)
        .then(isSupported => {
          if (typeof isSupported !== 'boolean') {
            throw new Error([
              'Invalid JS execution value returned from Sauce Labs.',
              `Received ${isSupported} and expected boolean`
            ].join((' ')));
          }
          return { record, isSupported, job };
        })
        .catch(console.log);
    })
  );

  await driver.quit();

  return items;
}

function log(browserName, version, platform, record: Object, isSupported: bool) {
  const shouldLogCompatSpecResults =
    process.env.LOG_COMPAT_SPEC_RESULTS
      ? process.env.LOG_COMPAT_SPEC_RESULTS === 'true'
      : true;

  if (shouldLogCompatSpecResults) {
    console.log([
      `"${record.protoChainId}" ${isSupported ? 'IS ✅ ' : 'is NOT ❌ '}`,
      `API supported in ${browserName} ${version} on ${platform}`
    ].join(' '));
  }
}

/**
 * Handle the respective results
 */
export async function handleFinishedTest(finishedTest: finishedTestType) {
  const { job, record, isSupported } = finishedTest;
  const { caniuseId, browserName, platform, version } = job;

  const existingRecordTargetVersions =
    await findSameVersionCompatRecord(record, caniuseId);

  // If newer version does not support API, current browser version doesn't support it
  // If older version does support API, current browser version does support it
  const { left, right } = getVersionsToMark(
    existingRecordTargetVersions
      ? Object.keys(existingRecordTargetVersions.versions)
      : [],
    caniuseId
  );

  // If middle is supported, mark all right as supported. Otherwise, mark
  // all left as unsupported
  await insertBulkRecords(
    record,
    caniuseId,
    [
      ...(isSupported ? right : left),
      String(version)
    ],
    isSupported
  );

  log(browserName, version, platform, record, isSupported);

  // Fetch the updated records from the TmpDatabase
  const newExistingRecordTargetVersions =
    await findSameVersionCompatRecord(record, caniuseId);

  // Create new jobs if there are more versions of the target to test
  const newVersions = getVersionsToMark(
    Object.keys(newExistingRecordTargetVersions.versions),
    caniuseId
  );
  if (newVersions.middle) {
    await jobQueue.insertBulk([{
      name: caniuseId,
      record: JSON.stringify(record),
      version: newVersions.middle,
      protoChainId: record.protoChainId,
      type: record.type,
      platform,
      browserName,
      caniuseId
    }]);
  }

  // Remove the finished job from the JobQueue
  return jobQueue.remove({
    name: caniuseId,
    protoChainId: record.protoChainId,
    type: record.type,
    record: JSON.stringify(record),
    version: String(version),
    platform,
    browserName,
    caniuseId
  });
}

type handleCapabilityType = Promise<Array<Promise<any>>>;

/**
 * @TODO: Optimizations:
 *          * Pass references to records instead of values. Possibly use ID in
 *            database. JSON.parse/stringify are expensive
 */
export async function handleCapability(capability: browserCapabilityType): handleCapabilityType {
  const { browserName, version } = capability;

  // Find all the jobs that match the current capability's browserName, version,
  // and platform
  const allJobs: Array<JobQueueType> = (await jobQueue.find({
    status: 'queued',
    browserName,
    version
  }))
  // @HACK: Temporarily avoid running on IE because of bugs with IE webdriver
  .filter(job => job.caniuseId !== 'ie');

  const jobs = allJobs.slice(
    parseInt(process.env.JOBS_INDEX_START, 10) || 0,
    parseInt(process.env.JOBS_INDEX_END, 10) || allJobs.length
  );

  await Promise.all(jobs.map(job => jobQueue.markJobsStatus(job, 'running'))); // eslint-disable-line

  return (await executeTests(capability, jobs)).map(handleFinishedTest);
}

/**
 * @NOTE: If you only want to test a few of these, remember to .slice(0, x) to
 *        test only the first x records. There's hundreds of records so this may
 *        take a while
 *
 * @TODO: Batch tests on browsers
 *
 * @TODO: Throttle promise
 */
async function Compat() {
  const browserCapabilities: browserCapabilityType[] = await setup();

  if (typeof browserCapabilities === 'boolean') {
    console.log('Jobs already exist in queue. No need to generate more');
    return process.exit(0);
  }

  // Dynamically generate compat-tests for each record and each browser
  await Promise.all(browserCapabilities.map(handleCapability));

  return process.exit(0);
}

export default Compat;

Compat();
