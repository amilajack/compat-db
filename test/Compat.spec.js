// @flow
import JobQueueDatabase from '../src/database/JobQueueDatabase';
import Compat, {
  executeTests,
  executeTestsParallel,
  handleFinishedTest,
  handleCapability
} from '../compat-tests/Compat';
import setup from '../compat-tests/setup';
import TmpRecordDatabase from '../src/database/TmpRecordDatabase';
import { baseRecord } from './JobQueueDatabase.spec';

jest.setTimeout(100000);

process.on('uncaughtException', err => {
  throw err;
});

const BROWSER_CAPABILITY = {
  browserName: 'chrome',
  version: '48.0',
  platform: 'Windows 10'
};

const TEST_RECORDS = [
  {
    protoChain: ['document', 'querySelector'],
    protoChainId: 'document.querySelector',
    name: 'document.querySelector',
    type: 'js-api'
  },
  {
    protoChain: ['CSSStyleDeclaration', 'borderWidth'],
    protoChainId: 'CSSStyleDeclaration.borderWidth',
    name: 'CSSStyleDeclaration.borderWidth',
    type: 'css-api'
  }
];

const [querySelectorRecord, borderWidthRecord] = TEST_RECORDS;

const TEST_JOBS = [
  {
    ...BROWSER_CAPABILITY,
    ...querySelectorRecord,
    record: JSON.stringify(querySelectorRecord),
    caniuseId: 'query-selector'
  },
  {
    ...BROWSER_CAPABILITY,
    ...borderWidthRecord,
    record: JSON.stringify(borderWidthRecord),
    caniuseId: 'border-width'
  }
];

const [querySelectorJob, borderWidthJob] = TEST_JOBS;

describe('Comapt', () => {
  beforeAll(() => {
    jest.spyOn(process, 'exit').mockImplementation(() => {});
  });

  it('should execute tests', async () => {
    expect(await executeTests(BROWSER_CAPABILITY, TEST_JOBS)).toEqual([
      {
        job: querySelectorJob,
        record: querySelectorRecord,
        isSupported: true
      },
      {
        job: borderWidthJob,
        record: borderWidthRecord,
        isSupported: true
      }
    ]);
  });

  it('should execute tests in parallel', async () => {
    expect(
      await executeTestsParallel(
        {
          browserName: 'chrome',
          version: '48.0',
          platform: 'Windows 10'
        },
        [true, false, true, true, false, true]
      )
    ).toEqual([true, false, true, true, false, true]);

    const exampleTest = `(function exampleTest() {
      return 'some';
    })()`;

    expect(
      await executeTestsParallel(
        {
          browserName: 'firefox',
          version: '42.0',
          platform: 'Windows 10'
        },
        [
          exampleTest,
          exampleTest,
          exampleTest,
          exampleTest,
          exampleTest,
          exampleTest
        ]
      )
    ).toEqual(['some', 'some', 'some', 'some', 'some', 'some']);
  });

  it('should handle finished tests', async () => {
    const jobQueue = new JobQueueDatabase('compat-test-1');
    const tmpRecordDatabase = new TmpRecordDatabase(
      'tmp-record-database-compat-1'
    );
    await jobQueue.migrate();
    await tmpRecordDatabase.migrate();

    // There should be no rows initially
    expect(await jobQueue.count()).toEqual(0);
    expect(await tmpRecordDatabase.count()).toEqual(0);

    // Manually insert initial jobs into job queue
    await jobQueue.connection.knex('compat-test-1').insert(
      TEST_JOBS.map(e => {
        delete e.protoChain;
        return e;
      })
    );
    expect(await jobQueue.count()).not.toEqual(0);

    const [finishedTest] = await executeTests(BROWSER_CAPABILITY, TEST_JOBS);
    expect(await jobQueue.count()).not.toEqual(0);
    await handleFinishedTest(
      finishedTest,
      'tmp-record-database-compat-1',
      jobQueue
    );
    expect(await jobQueue.count()).not.toEqual(0);

    // There should be one newly created record
    expect(await tmpRecordDatabase.count()).toEqual(1);
    await expect(tmpRecordDatabase.getAll()).resolves.toMatchSnapshot();

    // There should be one job left (because we created two jobs)
    expect(await jobQueue.count()).toEqual(1);
    await expect(tmpRecordDatabase.getAll()).resolves.toMatchSnapshot();
  });

  it('should handle capability', async () => {
    const jobQueue = new JobQueueDatabase('job-queue-test-handle-capability');
    await jobQueue.migrate();
    const result = await handleCapability(
      {
        browserName: 'chrome',
        version: '48.0',
        caniuseId: 'chrome',
        platform: 'Windows 10'
      },
      jobQueue
    );
    expect(result).toEqual([]);
  });

  it.skip('should run e2e', async () => {
    const jobQueue = new JobQueueDatabase('compat-test-e2e');
    const tmpRecordDatabase = new TmpRecordDatabase(
      'tmp-record-database-compat-2'
    );

    await jobQueue.migrate();

    // Initially, JobQueue should have no records
    expect(await jobQueue.count()).toEqual(0);
    expect(await tmpRecordDatabase.count()).toEqual(0);
    const firstRun = await jobQueue.getAll();

    await setup();

    // When running the jobs again, the jobs should not be the same
    await Compat();
    expect(await jobQueue.count()).toEqual(1);
    expect(await tmpRecordDatabase.count()).toEqual(1);
    const secondRun = await jobQueue.getAll();

    await Compat();
    expect(await jobQueue.count()).toEqual(1);
    expect(await tmpRecordDatabase.count()).toEqual(1);
    const thirdRun = await jobQueue.getAll();

    expect(firstRun).not.toEqual(secondRun);
    expect(thirdRun).not.toEqual(secondRun);
  });

  it.skip('should persist job on failure', async () => {
    const jobQueue = new JobQueueDatabase('job-queue-should-persist');
    expect(await jobQueue.count()).toEqual(0);

    await jobQueue.insertBulk([
      {
        ...baseRecord,
        version: '35.0',
        type: 'js-api',
        protoChainId: 'document.write',
        record: JSON.stringify({
          protoChain: ['document', 'write'],
          protoChainId: 'document.write',
          type: 'js-api'
        })
      }
    ]);

    expect(await jobQueue.getAll()).toEqual([
      {
        id: 1,
        name: 'chrome',
        browserName: 'chrome',
        platform: 'Windows 10',
        protoChainId: 'document.write',
        version: '35.0',
        type: 'js-api',
        record:
          '{"protoChain":["document","write"],"protoChainId":"document.write","type":"js-api"}',
        caniuseId: 'chrome',
        status: 'queued'
      }
    ]);

    expect(await jobQueue.count()).toEqual(1);

    await Compat();

    expect(await jobQueue.count()).toEqual(1);
  });

  // @TODO
  // it.skip('should handle capabilities', () => {});
});
