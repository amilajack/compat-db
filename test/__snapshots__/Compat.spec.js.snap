// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comapt should handle finished tests 1`] = `
Array [
  Object {
    "caniuseId": "query-selector",
    "id": 1,
    "name": "query-selector",
    "protoChainId": "document.querySelector",
    "type": "js-api",
    "versions": "{\\"48.0\\":\\"y\\"}",
  },
]
`;

exports[`Comapt should handle finished tests 2`] = `
Array [
  Object {
    "caniuseId": "query-selector",
    "id": 1,
    "name": "query-selector",
    "protoChainId": "document.querySelector",
    "type": "js-api",
    "versions": "{\\"48.0\\":\\"y\\"}",
  },
]
`;
